src/nvim/api/vim.c

```diff
@@ -2851,3 +2852,117 @@ void nvim_set_decoration_provider(Integer ns_id, DictionaryOf(LuaRef) opts,
 error:
   clear_provider(p);
 }
+
+///
+/// Set matches and start insert completion: see |complete()|.
+///
+/// @param startcol: Byte offset where the completed text starts.
+///                     Any text between startcol and the current
+///                     cursor col is replaced by the completion.
+/// @param matches: A list of string matches.
+/// @param opts: Dictionary. Possible keys:
+///                 filterfunc: function used for filtering the
+///                     matches. See |nvim_register_filterfunc|.
+//
+// todo: completefunc with refresh: always
+// todo: figure out what to do with ctrl_e and ctrl_l
+// todo: what to do if filterfunc != NULL AND completion entry
+// has icase/equal set.
+String nvim_complete(Integer startcol,
+                   Object matches,
+                   DictionaryOf(LuaRef) opts,
+                   Error *err)
+  FUNC_API_SINCE(7)
+{
+  typval_T tv;
+  extern LuaRef *user_filterfunc;
+  extern LuaRef *active_filterfunc;
+  LuaRef filterfunc;
+
+  if (matches.type != kObjectTypeArray) {
```

Comment:
do we intend `matches` to change type later? because otherwise we can just declare `Array` in the arg list

Reply:
I just took matches to be an object just for convenience, since we can then just use `object_to_vim` to convert to a vim list.

Reply:
you can do this anyway by using `ARRAY_OBJ` macro. Better to use `Array` in signature to get correct metadata.